swupd/packs_test.go

```diff
@@ -0,0 +1,401 @@
+package swupd
+
+import (
+	"fmt"
+	"io"
+	"os"
+	"reflect"
+	"sort"
+	"strings"
+	"testing"
+)
+
+func TestFindBundlesToPack(t *testing.T) {
+	type M map[string]uint32
+	tests := []struct {
+		Name       string
+		FromV      uint32
+		From       M
+		ToV        uint32
+		To         M
+		Expected   []BundleToPack
+		ShouldFail bool
+	}{
+		{
+			Name: "simple case",
+
+			FromV: 10,
+			From:  M{"os-core": 10},
+
+			ToV: 20,
+			To:  M{"os-core": 20},
+
+			Expected: []BundleToPack{{"os-core", 10, 20}},
+		},
+
+		{
+			Name: "from 0 keeps all bundles",
+
+			ToV: 10,
+			To:  M{"os-core": 10, "editors": 10},
+
+			Expected: []BundleToPack{{"os-core", 0, 10}, {"editors", 0, 10}},
+		},
+
+		{
+			Name: "new manifest has bundle not present in the old",
+
+			FromV: 10,
+			From:  M{"os-core": 10},
+
+			ToV: 20,
+			To:  M{"os-core": 20, "manuals": 20},
+
+			Expected: []BundleToPack{{"os-core", 10, 20}, {"manuals", 0, 20}},
+		},
+
+		{
+			Name: "new manifest has bundles from old",
+
+			FromV: 10,
+			From:  M{"os-core": 10, "editors": 10, "c-basic": 10, "R-basic": 10},
+
+			ToV: 20,
+			To:  M{"os-core": 20, "editors": 10, "c-basic": 20, "R-basic": 10},
+
+			Expected: []BundleToPack{{"os-core", 10, 20}, {"c-basic", 10, 20}},
+		},
+
+		{
+			Name: "both using an older bundle",
+
+			FromV: 100,
+			From:  M{"os-core": 100, "editors": 20},
+
+			ToV: 200,
+			To:  M{"os-core": 200, "editors": 20},
+
+			Expected: []BundleToPack{{"os-core", 100, 200}},
+		},
+
+		{
+			Name: "bundles with versions that dont match the manifest",
+
+			FromV: 100,
+			From:  M{"os-core": 100, "c-basic": 80},
+
+			ToV: 200,
+			To:  M{"os-core": 200, "c-basic": 150},
+
+			Expected: []BundleToPack{{"os-core", 100, 200}, {"c-basic", 80, 150}},
+		},
+
+		{
+			Name: "old bundle not around anymore",
+
+			FromV: 100,
+			From:  M{"os-core": 100, "b-basic": 100},
+
+			ToV: 200,
+			To:  M{"os-core": 200, "c-basic": 200},
+
+			Expected: []BundleToPack{{"os-core", 100, 200}, {"c-basic", 0, 200}},
+		},
+	}
+
+	addBundle := func(m *Manifest, name string, version uint32) {
+		bundle := &File{
+			Name:    name,
+			Type:    TypeManifest,
+			Version: version,
+		}
+		m.Files = append(m.Files, bundle)
+	}
+
+	sortAndPrintBundles := func(bundles []BundleToPack) {
+		sort.Slice(bundles, func(i, j int) bool {
+			return bundles[i].Name < bundles[j].Name
+		})
+		for _, b := range bundles {
+			fmt.Printf("  %s %d -> %d\n", b.Name, b.FromVersion, b.ToVersion)
+		}
+		fmt.Println()
+	}
+
+	for _, tt := range tests {
+		var fromM *Manifest
+		if tt.FromV != 0 {
+			fromM = &Manifest{}
+			fromM.Header.Version = tt.FromV
+			for name, v := range tt.From {
+				addBundle(fromM, name, v)
+			}
+		}
+		toM := &Manifest{}
+		toM.Header.Version = tt.ToV
+		for name, v := range tt.To {
+			addBundle(toM, name, v)
+		}
+
+		bundleMap, err := FindBundlesToPack(fromM, toM)
+		if err != nil {
```

Comment:
I've suggested this elsewhere, you can (optionally) simplify this logic by doing something like
```go
if (err != nil) != tt.ShouldFail {
    if tt.ShouldFail {
        t.Fatalf(...)
    } else {
        t.Fatalf(...)
    }
    continue
}
```

Reply:
I haven't looked at the BCB review, but I will try that approach there.

Reply:
It did look better on the BCB, so I'll try the same thing here.

Reply:
Done.